lib/TokenManager.js

```diff
@@ -91,18 +34,36 @@ function add(sdk, tokenMgmtRef, storage, key, token) {
   }
   tokenStorage[key] = token;
   storage.setStorage(tokenStorage);
-  setRefreshTimeout(sdk, tokenMgmtRef, storage, key, token);
 }
 
 function get(storage, key) {
   var tokenStorage = storage.getStorage();
   return tokenStorage[key];
 }
 
-function remove(tokenMgmtRef, storage, key) {
-  // Clear any listener for this token
-  clearRefreshTimeout(tokenMgmtRef, key);
+function getAsync(sdk, tokenMgmtRef, storage, key) {
+  var token = get(storage, key);
+  var deferred = Q.defer();
 
+  if (token && (token.expiresAt * 1000 <= Date.now())) {
```

Comment:
Shouldn't this be `>=`?  Tell me if I haven't had enough ☕️ yet

I'm noticing that a problem is still present: if the client clock is not in sync with the issuer clock, our "is expired" check can be incorrect.  One idea for solving this: when storing a new token, take the difference between `iat` and `exp`, then add that to the current client time.  Store the resulting value as the client time at which the token is considered expired.

Reply:
This logic is correct - but comparing dates trips me up sometimes as well.

The `token.expiresAt * 1000` value is giving us the time the token has expired. This value **MUST** be less than the current date. For example, if the token expires at `5:00pm`, and the current time is `5:30pm`, the token is no longer valid.

Reply:
The problem you're referencing is partially solved with `clockskew`. We have some information about that [here](https://github.com/okta/okta-auth-js/blob/master/lib/clientBuilder.js#L61-L67).

This is a common problem throughout the industry, where **client** and **server** times are oftentimes out of sync (mostly seen with daylight savings). I've seen the following alternatives:

- Implement a "system time" API endpoint on the server
- Validate the token against the `/introspection` endpoint

Reply:
All good points.  My thinking was proactive about some cases that seem to be edge (since we haven't heard about them) where the clock may be significantly off.  No changes needed.